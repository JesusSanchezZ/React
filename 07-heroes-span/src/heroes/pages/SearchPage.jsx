import { useNavigate, useLocation } from 'react-router-dom';
import queryString from 'query-string';
import { useForm } from '../../hooks/useForm';
import { HeroCard } from '../components'
import { getHeroesByName } from '../helpers';

export const SearchPage = () => {
    const navigate = useNavigate();             // obtenemos la navegaci贸n 
    const location = useLocation();             // obtener informaci贸n de la ubicaci贸n donde nos encontramos

    //const query = queryString.parse(location.search);
    //console.log({query});
    const { q = ''} = queryString.parse(location.search);

    const heroes = getHeroesByName(q);

    const showSearch = (q.length === 0);
    const showError = (q.length > 0) && (heroes.length === 0);

    const {searchText, onInputChange} = useForm({
        searchText: q
    });

    const onSearchSubmit = (event) => {
        event.preventDefault();
        //if( searchText.trim().length <= 1 ) return;

        navigate(`?q=${ searchText }`);
        //console.log({ searchText });
    }

    return (
        <>
            <h1>Search</h1>
            <hr />

            <div className="row">
                <div className="col-5">
                    <h4>Searching</h4>
                    <hr />
                    <form 
                        aria-label='form'
                        onSubmit={ onSearchSubmit }
                    >
                        <input
                            type="text"
                            placeholder="Search a hero"
                            className="form-control"
                            name="searchText"
                            autoComplete="off"
                            value={ searchText }
                            onChange={ onInputChange }
                        />
                        <button className="btn btn-outline-primary form-control mt-1">
                            Search
                        </button>
                    </form>
                </div>

                <div className="col-7">
                    <h4><span className='badge rounded-pill bg-dark'>{heroes.length}</span> Results</h4>
                    <hr />

                    {/* {
                        ( q === '')
                        ? <div className="alert alert-primary">Search a hero</div>
                        : ( heroes.length === 0)
                            && <div className="alert alert-danger">No hero with <b>{q}</b></div>
                    } */}

                    <div 
                        aria-label="search-hero"
                        className="alert alert-primary animate__animated animate__rubberBand" 
                        style={{ display: showSearch ?'': 'none'}}
                    >
                        Search a hero
                    </div>

                    <div 
                        aria-label="alert-danger"
                        className="alert alert-danger animate__animated animate__rubberBand" 
                        style={{ display: showError ?'': 'none'}}
                    >
                        No hero with <b>{q}</b>
                    </div>

                    {
                        heroes.map( hero => (
                            <HeroCard key={hero.id} {...hero} />
                        ))
                    }
                </div>
            </div>
        </>
    );
}


/* 
    Instalaci贸n de paquete query string para leer propiedades del url
        npm install query-string
    */