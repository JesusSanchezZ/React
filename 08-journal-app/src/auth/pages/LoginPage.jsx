import { Link as RouterLink } from "react-router-dom";
import { Google } from "@mui/icons-material";
import { Alert, Grid, TextField, Typography, Button, Link } from "@mui/material";
import { AuthLayout } from "../layout/AuthLayout";
import { useForm } from "../../hooks";
import { useMemo, useState } from "react";
import { startGoogleSignIn, startLoginWithEmailPassword } from "../../store/auth";
import { useDispatch, useSelector } from "react-redux";

const formValidations = {
    email: [ (value) => value.length >=1, 'El correo es necesario'],
    password: [ (value) => value.length >= 1, 'El password es necesario']
}

const formData = {
    email: 'jesus@sanchez.com',
    password: '123456'
}

export const LoginPage = () =>{
    const { status, errorMessage } = useSelector( state => state.auth );
    const [formSubmitted, setFormSubmitted] = useState(false);

    const dispatch = useDispatch();

    const { 
        email, password, onInputChange, formState,
        isFormValid, emailValid, passwordValid 
    } = useForm(formData, formValidations);

    const isAuthenticated = useMemo( () => status === 'checking', [status]);

    // useEffect(() => {
    //     dispatch( checkingAuthentication() );
    // },[]);

    const onSubmit = ( event ) => {
        event.preventDefault();
        setFormSubmitted(true);

        // dispatch( checkingAuthentication() );
        // console.log({ email, password });
        if(!isFormValid) return;
        dispatch( startLoginWithEmailPassword({email, password}) );
    }

    const onGoogleSignIn = () => {
        dispatch( startGoogleSignIn() );
        console.log('onGoogleSignIn');
    }

    return (
        <AuthLayout title="Login">
            <form 
                className='animate__animated animate__fadeIn animate__faster'
                onSubmit={ onSubmit }
            >
                <Grid container>
                    <Grid item xs={12} sx={{ mt: 2 }}>
                        <TextField
                            label="Correo"
                            type="email"
                            placeholder="correo@google.com"
                            fullWidth
                            name="email"
                            value={ email }
                            onChange={ onInputChange }
                            error={ !!emailValid && formSubmitted }
                            helperText={emailValid}
                        />
                    </Grid>

                    <Grid item xs={12} sx={{ mt: 2 }}>
                        <TextField
                            label="Contraseña"
                            type="password"
                            placeholder="Contraseña"
                            fullWidth
                            name="password"
                            value={ password }
                            onChange={ onInputChange }
                            error={ !!passwordValid && formSubmitted }
                            helperText={ passwordValid }
                        />
                    </Grid>

                    <Grid container spacing={2} sx={{ mb: 2, mt: 1 }}>
                        <Grid
                            item
                            xs={12}
                            display={ !!errorMessage ? '':'none'}
                        >
                            <Alert severity='error'>{ errorMessage }</Alert>
                        </Grid>
                        <Grid item xs={12} sm={6}>
                            <Button 
                                disabled={ isAuthenticated }
                                type="submit" 
                                variant="contained" 
                                fullWidth
                            >
                                Login
                            </Button>
                        </Grid>
                        <Grid item xs={12} sm={6}>
                            <Button 
                                disabled={ isAuthenticated }
                                onClick={ onGoogleSignIn } 
                                variant="contained" 
                                fullWidth
                            >
                                <Google />
                                <Typography sx={{ ml: 1 }}>Google</Typography>
                            </Button>
                        </Grid>
                    </Grid>

                    <Grid container direction='row' justifyContent='end'>
                        <Link component={RouterLink} color="inherit" to="/auth/register">
                            Crear una cuenta
                        </Link>
                    </Grid>

                </Grid>
            </form>
        </AuthLayout>
    );
}