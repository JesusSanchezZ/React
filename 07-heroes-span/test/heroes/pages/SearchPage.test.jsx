import { fireEvent, render, screen } from '@testing-library/react';
import { unstable_renderSubtreeIntoContainer } from 'react-dom';
import { createRenderer } from 'react-dom/test-utils';
import { MemoryRouter, useNavigate } from 'react-router-dom';
import { SearchPage } from '../../../src/heroes/pages/SearchPage';

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom',() => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUseNavigate,
}));

describe('Pruebas en <SearchPage/>', ()=> {
    beforeEach(() => jest.clearAllMocks());

    test('Debe de mostrar correctamente con valores por defecto',()=>{
        const { container } = render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        );

        //expect( container ).toMatchSnapshot();
        //screen.debug();
    });

    test('Debe de mostrar a Batman y el input con el valor del queryString',()=>{
        const { container } = render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchPage />
            </MemoryRouter>
        );

        //screen.debug();
        const input = screen.getByRole('textbox');
        expect( input.value ).toBe('batman');

        const img = screen.getByRole('img');
        expect( img.src ).toContain('/assets/heroes/dc-batman.jpg');

        const searchDiv = screen.getByLabelText('search-hero');
        expect( searchDiv.style.display ).toBe('none');
    });

    test('Debe de mostrar un error si no se encuenctra el heroe',()=>{
        render(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <SearchPage />
            </MemoryRouter>
        );

        const alert = screen.getByLabelText('alert-danger');
        
        //screen.debug();
        expect( alert.style.display).toBe('');
    });

    test('Debe de llamar el navigate a la pantalla nueva',()=>{
        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchPage />
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        fireEvent.change( input, { target: {name: 'searchText', value: 'superman'}});

        const form = screen.getByRole('form');
        fireEvent.submit(form);

        expect( mockedUseNavigate ).toHaveBeenCalledWith('?q=superman');
    });
});