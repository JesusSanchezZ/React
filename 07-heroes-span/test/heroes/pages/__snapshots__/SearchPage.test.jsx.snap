// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPage/> Debe de mostrar correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form>
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary form-control mt-1"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        <span
          class="badge rounded-pill bg-dark"
        >
          0
        </span>
         Results
      </h4>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__rubberBand"
      >
        Search a hero
      </div>
      <div
        class="alert alert-danger animate__animated animate__rubberBand"
        style="display: none;"
      >
        No hero with 
        <b />
      </div>
    </div>
  </div>
</div>
`;
